app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx

```diff
@@ -0,0 +1,278 @@
+  );
+};
+
+/**
+ * A component for dynamically searching for users and displaying results in a dropdown.
+ * This component will replace the textarea passed in props, copying all styles and attributes, and allowing for progressive enhancement
+ *
+ * @param {object} props
+ * @param {element} props.replaceElement The textarea DOM element that should be replaced
+ * @param {function} props.fetchSuggestions The async call to use for the search
+ *
+ * @example
+ * <MentionAutocompleteCombobox
+ *    replaceElement={textAreaRef.current}
+ *    fetchSuggestions={fetchUsersByUsername}
+ * />
+ */
+export const MentionAutocompleteTextArea = ({
+  replaceElement,
+  fetchSuggestions,
+}) => {
+  const [textContent, setTextContent] = useState('');
+  const [searchTerm, setSearchTerm] = useState('');
+  const [cachedSearches, setCachedSearches] = useState({});
+  const [dropdownPositionPoints, setDropdownPositionPoints] = useState({
+    x: 0,
+    y: 0,
+  });
+  const [selectionInsertIndex, setSelectionInsertIndex] = useState(0);
+  const [users, setUsers] = useState([]);
+  const [cursorPosition, setCursorPosition] = useState(null);
+  const [ariaHelperText, setAriaHelperText] = useState('');
+
+  const isSmallScreen = useMediaQuery(`(max-width: ${BREAKPOINTS.Small}px)`);
+
+  const inputRef = useRef(null);
+
+  useEffect(() => {
+    if (searchTerm.length >= MIN_SEARCH_CHARACTERS) {
+      if (cachedSearches[searchTerm]) {
+        setUsers(cachedSearches[searchTerm]);
+        return;
+      }
+
+      fetchSuggestions(searchTerm).then(({ result: fetchedUsers }) => {
+        const resultLength = Math.min(
+          fetchedUsers.length,
+          MAX_RESULTS_DISPLAYED,
+        );
+
+        const results = fetchedUsers.slice(0, resultLength);
+
+        setCachedSearches({
+          ...cachedSearches,
+          [searchTerm]: results,
+        });
+
+        setUsers(results);
+
+        // Let screen reader users know a list has populated
+        const requiresAriaLiveAnnouncement =
+          !ariaHelperText && fetchedUsers.length > 0;
+
+        if (requiresAriaLiveAnnouncement) {
+          setAriaHelperText(
+            `Mention user, ${fetchedUsers.length} results found`,
+          );
+        }
+      });
+    }
+  }, [searchTerm, fetchSuggestions, cachedSearches, ariaHelperText]);
+
+  useLayoutEffect(() => {
+    const popover = document.getElementById('mention-autocomplete-popover');
+    if (!popover) {
+      return;
+    }
+
+    const closeOnClickOutsideListener = (event) => {
+      if (!popover.contains(event.target)) {
+        // User clicked outside, reset to not searching state
+        setSearchTerm('');
+        setAriaHelperText('');
+        setUsers([]);
+      }
+    };
+
+    document.addEventListener('click', closeOnClickOutsideListener);
+
+    return () =>
+      document.removeEventListener('click', closeOnClickOutsideListener);
+  }, [searchTerm]);
+
+  useLayoutEffect(() => {
+    inputRef.current.focus();
```

Comment:
Refactor (blocking): Create a variable to reference `inputRef.current` to avoid looking up the `current` property all the time.